### PR TITLE
Add mediation to credential creation options

### LINKED ISSUE #225: Mediation parameter for credential creation
It is desirable to be able to specify a CredentialMediationRequirement parameter for conditional creation, similar to how it's available for get.

This will be useful for RPs to specify that they desire credential creation, but do not a modal flow for it. This would be useful for conditional registration in Web Authentication.

### PR DESCRIPTION
**Pull request by pascoej**
This fixes 

Most of the heavy lifting of conditional in create is handled in WebAuthn.

Fixes #225

See also:
https://github.com/w3c/webauthn/pull/1951/files
https://github.com/w3c/webauthn/wiki/Explainer:-Conditional-Registration-Extension

***
<a href="https://pr-preview.s3.amazonaws.com/pascoej/webappsec-credential-management/pull/224.html" title="Last updated on May 30, 2024, 5:14 PM UTC (ffa7bbc)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webappsec-credential-management/224/a70bc66...pascoej:ffa7bbc.html" title="Last updated on May 30, 2024, 5:14 PM UTC (ffa7bbc)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -790,6 +790,11 @@ spec:css-syntax-3;
         interfaces it refers to=] have overridden {{Credential/isConditionalMediationAvailable()}} to return
         a new {{Promise}} that [=resolves=] with `true`.
 
+        For {CredentialsContainer/create()}, if a user has previously consented to credential creation and
+        the user agent knows it recently mediated an authentication, then the `create()` call may resolve without
+        additional prominent modal interaction. If the user agent did not recently mediate an authentication or
+        does not have consent for credential creation, then the call must throw a "{{NotAllowedError}}" {{DOMException}}.
```
Comment (nsatragno) [inline index.bs:796]: > the call must throw a "{{NotAllowedError}}" {{DOMException}}.

We should spell this out on the `Create a credential` algorithm below.

**Review by nsatragno — CHANGES_REQUESTED** [review]
Thank you for your CL! The API surface looks good on the credman side. We just have to make sure the interpretation of mediation remains clear for other credential types. Here is my first pass.

**Review by Hmercer23 — APPROVED** [review]

File: index.bs [hunk 2]
```diff
@@ -1230,6 +1086,175 @@ spec:css-syntax-3;
...
+        be. If [=requires user mediation|user mediation is required=], then the user agent will
+        involve the user in the decision.
+
+        Note: This is the default behavior for {{get()}} and {{create()}}, and is intended to serve a case where a
+        developer has reasonable confidence that a user expects to start an operation. If
+        a user has just clicked "sign-in" for example, then they won't be surprised or confused to
+        see a [=credential chooser=] if necessary.
+
+    :   <dfn>conditional</dfn>
+    ::  For {CredentialsContainer/get()}, discovered credentials are presented to the user in a non-modal dialog along with an
```
Comment (nsatragno) [inline index.bs:1127]:
```suggestion
    ::  For {{CredentialsContainer/get()}}, discovered credentials are presented to the user in a non-modal dialog along with an
```

File: index.bs [hunk 3]
```diff
@@ -1230,6 +1086,175 @@ spec:css-syntax-3;
...
+        that depends on the type of credential (e.g. to insert a device containing credentials).
+        Either way, the `get()` method MUST NOT resolve immediately with `null` to avoid revealing
+        the lack of applicable credentials to the website.
+
+        Websites can only pass {{CredentialMediationRequirement/conditional}} into the
+        {{CredentialsContainer/get()}} method if all of the [=relevant credential interface objects|credential
+        interfaces it refers to=] have overridden {{Credential/isConditionalMediationAvailable()}} to return
+        a new {{Promise}} that [=resolves=] with `true`.
+
+        For {CredentialsContainer/create()}, if a user has previously consented to credential creation and
```
Comment (nsatragno) [inline index.bs:1146]:
```suggestion
        For {{CredentialsContainer/create()}}, if a user has previously consented to credential creation and
```

File: index.bs [hunk 4]
```diff
@@ -1230,6 +1086,175 @@ spec:css-syntax-3;
...
+        the user agent knows it recently mediated an authentication, then the `create()` call may resolve without
+        additional prominent modal interaction. If the user agent did not recently mediate an authentication or
+        does not have consent for credential creation, then the call must throw a "{{NotAllowedError}}" {{DOMException}}.
+
+        Websites can only pass {{CredentialMediationRequirement/conditional}} into the
+        {{CredentialsContainer/create()}} method if |conditionalCreate| is present in the result of
+        {{PublicKeyCredential.getClientCapabilities()}} defined in [[WEBAUTHN]].
```
Comment (nsatragno) [inline index.bs:1153]: This paragraph implies that conditional creation is supported for all credential types if |conditionalCreate| is present in `getClientCapabilities()`. Let's make clear it's only available for webauthn credentials:

```suggestion
        Websites can only pass {{CredentialMediationRequirement/conditional}} into the
        {{CredentialsContainer/create()}} method for {{PublicKeyCredential}} requests, and
        only if {{ClientCapability/conditionalCreate}} is present in the result of
        {{PublicKeyCredential.getClientCapabilities()}} as defined in [[WEBAUTHN]].
```

I'm not too happy that this webauthn-specific detail is leaking into credman, but it's probably okay to live with it for now. If in the future more credential types wanted to support this, we could make a more generic interface to feature detect.

File: index.bs [hunk 5]
```diff
@@ -913,6 +754,13 @@ spec:css-syntax-3;
     ::  This property lets the developer abort an ongoing {{CredentialsContainer/create()}}
         operation. An aborted operation may complete normally (generally if the abort was received
         after the operation finished) or reject with an [=AbortSignal/abort reason=].
+
+    :   <dfn>mediation</dfn>
+    ::  Certain [=credential type registry/credential types=] may require [=user mediation=] before creating
+        a {{Credential}} via {{CredentialsContainer/create()}}, e.g. if doing so nessessitates storing it on
+        an external device or requesting information from a third party. These [=credential types=] can be
+        specialized to talior said [=user mediation=] depending on the value of |mediation|, as described below
```
Comment (nsatragno) [inline index.bs:762]:
```suggestion
        specialized to tailor said [=user mediation=] depending on the value of |mediation|, as described below
```

**Review by nsatragno — APPROVED** [review]
Approved with comments. Thank you for your great work!

**Comment by pascoej** [comment]
Thank you for the review!

File: index.bs [hunk 6]
```diff
@@ -434,6 +435,20 @@ spec:css-syntax-3;
...
+        to create a credential using the {{CredentialMediationRequirement/conditional}} approach to
+        [[#mediation-requirements|mediation of credential creation]] for the [=credential/credential type=].
+
+        {{Credential}}'s default implementation of {{Credential/willRequestConditionalCreation()}}:
+
+        <ol class="algorithm">
+            1.  Return a new {{Promise}} that immediately [=resolves=].
+        </ol>
+
+        Note: If this function is not present, {{CredentialMediationRequirement/conditional}}
```
Comment (marcoscaceres) [inline index.bs:449]: This should probably be "method" or "operation"

```suggestion
        Note: If this method is not present, {{CredentialMediationRequirement/conditional}}
```

File: index.bs (hunk 1 above)
Comment (marcoscaceres) [reply index.bs:811]: Maybe this should actually say, "will return [=a promise rejected with=] a "{{NotAllowedError}}" {{DOMException}}", as the method itself doesn't "throw", it returns a promise.

**Review by marcoscaceres — COMMENTED** [review]
Noticed a couple small issues here... might be good to fix up before merging.

**Pull request merged**